### PR TITLE
Build Horovod with temporarily installed CMake if necessary

### PR DESCRIPTION
**Pull request by maxhgerlach**
## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs?
- [x] Did you write any tests to validate this change?  
- [ ] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

This is a follow-up to PR #3261, which would require users of many distros to install a recent CMake from an external source (like `pip`).

If such a situation is identified in `setup.py`, we can automatically install a sufficiently recent CMake to a temporary directory and use that to build Horovod. That directory is cleaned up afterwards and the user or system environment is not affected. If `HOROVOD_CMAKE` is set, only that CMake binary is used without any change of behavior.

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

File: Dockerfile.test.cpu
Comment (EnricoMi) [inline Dockerfile.test.cpu:36]: we could rely here on `setup.py` installing the latest cmake so we always test this really works

File: .github/workflows/codeql-analysis.yml
Comment (EnricoMi) [inline .github/workflows/codeql-analysis.yml:103]: Why is installing `cmake` through `apt-get install`in the `Dockerfile`s not sufficient here?

File: Dockerfile.test.cpu
Comment (maxhgerlach) [reply Dockerfile.test.cpu:36]: I believe that this apt-get line currently installs cmake 3.10 on the Ubuntu 18.04 we use for testing (too old to build Horovod). So yes, as it stands, we rely on setup.py to install a temporary CMake 3.13 to build Horovod and the mechanism from this PR is tested. This would change once we update to Ubuntu 20.04 or later. 

I left cmake in here for any non-Horovod builds in this Docker container, e.g., for oneCCL.

File: .github/workflows/codeql-analysis.yml
Comment (maxhgerlach) [reply .github/workflows/codeql-analysis.yml:103]: We need CMake 3.13 to compile the C++ parts of Horovod. On Ubuntu 18.04 and without extra package sources like a PPA from Kitware `apt-get install cmake` installs CMake 3.10 IIRC.

The mechanism of this PR does not work with the build for CodeQL because we don't use setup.py there.

File: Dockerfile.test.cpu
Comment (EnricoMi) [reply Dockerfile.test.cpu:36]: Ah, our ancient Ubuntu 18.04 means this `apt-get intall cmake` is not sufficient. Then it makes sense we need to add that extra code to the CodeQL yaml. Once we update to Ubuntu 20.04 we will lose testing that auto-install. We should remember when we upgrade to keep Ubuntu 18.04 for one of the test combinations, e.g. TF 1.15 to see that auto-install feature still being tested (as long as we want to support Ubuntu 18.04 in our test setup).

**Review by EnricoMi — COMMENTED** [review]
I link this, leave it to @tgaddair and @nvcastet to approve.

File: Dockerfile.test.cpu
Comment (maxhgerlach) [reply Dockerfile.test.cpu:36]: > We should remember when we upgrade to keep Ubuntu 18.04 for one of the test combinations, e.g. TF 1.15 to see that auto-install feature still being tested (as long as we want to support Ubuntu 18.04 in our test setup).

That's a good plan!

> https://github.com/horovod/horovod/runs/4852098212?check_suite_focus=true#step:10:4652

Yep, here we can see that it used a CMake binary from `/tmp`:

>   11.63   Could not find a recent CMake to build Horovod. **Attempting to install CMake 3.13 to a temporary location via pip.**
>   ...
>   13.92   Running CMake in build/temp.linux-x86_64-3.7/RelWithDebInfo:
>   13.92   **/tmp/horovod-cmake-tmpmj4n5l40/bin/run_cmake** /tmp/pip-req-build-fzb3ontk -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_LIBRARY_OUTPUT_DIRECTORY_RELWITHDEBINFO=/tmp/pip-req-build-fzb3ontk/build/lib.linux-x86_64-3.7 -DPYTHON_EXECUTABLE:FILEPATH=/usr/bin/python
>

File: .github/workflows/codeql-analysis.yml
Comment (EnricoMi) [reply .github/workflows/codeql-analysis.yml:103]: This can probably be removed again as we have moved to Ubuntu 20.04 for our test images in the meantime.

File: Dockerfile.test.cpu
Comment (EnricoMi) [reply Dockerfile.test.cpu:36]: As a side note: In the meantime we have moved our test images to Ubuntu 20.04, but a few test images still use Ubuntu 18.04, so we now test both.

**Review by EnricoMi — CHANGES_REQUESTED** [review]

File: .github/workflows/codeql-analysis.yml
Comment (maxhgerlach) [reply .github/workflows/codeql-analysis.yml]: I removed this and also went back to installing CMake via apt in the Dockerfiles based on Ubuntu 20.04 (`docker/*/Dockerfile`).

**Review by EnricoMi — APPROVED** [review]
LGTM!

**Review by nvcastet — APPROVED** [review]

**Comment by EnricoMi** [comment]
@maxhgerlach I have rebased this with master, lets see if we now get the tests terminate

**Comment by EnricoMi** [comment]
@maxhgerlach looks like the last minute change to the `horovod` and `horovod-cpu` images broke Docker image generation. The GitHub action steps take forever (what I observed in this branch before merge, and which is the case for branches rebased to master since):

https://github.com/horovod/horovod/runs/5385819691?check_suite_focus=true
https://github.com/horovod/horovod/runs/5385981381?check_suite_focus=true

**Comment by maxhgerlach** [comment]
Great find, thanks @EnricoMi!

Somehow Github Actions doesn't show me any logs for the hanging steps. Is there any way to get them?

**Comment by EnricoMi** [comment]
Note to self: add timeout to docker build jobs ;-)

**Comment by maxhgerlach** [comment]
> While a job is running, only new lines are shown.
And the only way to get to see those appears to be to keep a tab open and to not let your computer go to sleep...

**Pull request merged**
